### PR TITLE
Make the AST the canonical source for window frame information

### PR DESCRIPTION
**Pull request by ebd2**
The AST currently contains empty window frames and frame end bounds when none are specified. Later the planner creates WindowNodes containing window frames (and frame end bounds) with the SQL standard defaults when the AST has empty frames (or bounds).

This has two effects:
1) the AST is not the canonical source for frame information
2) The AST and the plan are inconsistent in terms of what they contain

Ensuring that the AST contains the appropriate defaults resolves both of those and also simplifies a whole bunch of places in the analyzer and planner that currently have to handle the default window frame as a special case.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -500,23 +500,21 @@ private PlanBuilder window(PlanBuilder subPlan, QuerySpecification node)
...
+            WindowFrame.Type frameType = null;
+            FrameBound.Type frameStartType = null;
+            FrameBound.Type frameEndType = null;
             Expression frameStart = null;
             Expression frameEnd = null;
 
-            if (window.getFrame().isPresent()) {
-                WindowFrame frame = window.getFrame().get();
-                frameType = frame.getType();
+            WindowFrame frame = window.getFrame();
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Move it before https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#diff-21d31cb9d487772413eff4f67b8555deR503 to avoid unnecessary null-initialization of e.g. `frameType`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -500,23 +500,21 @@ private PlanBuilder window(PlanBuilder subPlan, QuerySpecification node)
...
+            frameType = frame.getType();
 
-                frameStartType = frame.getStart().getType();
-                frameStart = frame.getStart().getValue().orElse(null);
+            frameStartType = frame.getStart().getType();
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Initialize local variable earlier (like https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#r74996131).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 3]
```diff
@@ -557,7 +555,7 @@ private PlanBuilder window(PlanBuilder subPlan, QuerySpecification node)
                 frameEndSymbol = Optional.of(subPlan.translate(frameEnd));
             }
 
-            WindowNode.Frame frame = new WindowNode.Frame(frameType,
+            WindowNode.Frame rewrittenFrame = new WindowNode.Frame(frameType,
                     frameStartType, frameStartSymbol,
                     frameEndType, frameEndSymbol);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java:550]: Is it really equivalent to what was before?
When there's no `end` given for a frame, then https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#diff-21d31cb9d487772413eff4f67b8555deL516 does not set `frameEndType`, nor `frameEnd` which falls back to default of https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#diff-21d31cb9d487772413eff4f67b8555deL505.
On this patch if we won't enter https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#diff-21d31cb9d487772413eff4f67b8555deR515 we end up setting default value - `null` (https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#diff-21d31cb9d487772413eff4f67b8555deR505).
Is that correct behavior?

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 4]
```diff
@@ -1363,12 +1363,12 @@ protected Node aggregateResult(Node aggregate, Node nextResult)
         throw new UnsupportedOperationException("not yet implemented");
     }
 
-    private <T> Optional<T> visitOrElse(ParserRuleContext context, Class<T> clazz, T other)
+    private <T> T visitOrElse(ParserRuleContext context, Class<T> clazz, T other)
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Unrelated change. Maybe worth extracting to separate commit?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ExpressionTreeRewriter.java [hunk 5]
```diff
@@ -500,37 +500,35 @@ public Expression visitFunctionCall(FunctionCall node, Context<C> context)
...
-                    WindowFrame frame = rewrittenFrame.get();
-
-                    FrameBound start = frame.getStart();
-                    if (start.getValue().isPresent()) {
-                        Expression value = rewrite(start.getValue().get(), context.get());
-                        if (value != start.getValue().get()) {
-                            start = new FrameBound(start.getType(), value);
-                        }
+                WindowFrame rewrittenFrame = window.getFrame();
+                WindowFrame frame = rewrittenFrame;
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ExpressionTreeRewriter.java]: Unnecessary local variable.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 6]
```diff
@@ -500,22 +500,20 @@ private PlanBuilder window(PlanBuilder subPlan, QuerySpecification node)
             Window window = windowFunction.getWindow().get();
 
             // Extract frame
-            WindowFrame.Type frameType = null;
-            FrameBound.Type frameStartType = null;
-            FrameBound.Type frameEndType = null;
-            Expression frameStart = null;
-            Expression frameEnd = null;
+            WindowFrame.Type frameType;
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Unnecessary change (should be squashed with previous commit). But if you apply my suggestion (from comment https://github.com/Teradata/presto/pull/304/commits/f5a57febdbef7b62f1ef628aaede4c40d666c32b#r74996131) it will be irrelevant anyway.

**Comment by maciejgrzybek** [comment]
I'd consider moving a60bf5c294f18e679db7b099c3a3aa7f9f21abec before f5a57febdbef7b62f1ef628aaede4c40d666c32b. Would look more consistent. But I won't insist.

**Comment by maciejgrzybek** [comment]
Back to you @ebd2.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 3 above)
Comment (ebd2) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java:550]: Good catch. Thanks.

**Comment by ebd2** [comment]
I think I've covered everything. Back to you @maciejgrzybek

**Comment by maciejgrzybek** [comment]
Since `FrameBound` implicitly assume value have to be `Optional.empty()` if type = `UNBOUNDED_PRECEDING` || `UNBOUNDED_FOLLOWING`, I think it would be valuable to add such a assertion in `FrameBound` c-tor. Though, it's not related change to what you've done but you play with this part of code already, so probably it's worth to address that with this PR.

**Comment by ebd2** [comment]
https://github.com/Teradata/presto/pull/304#issuecomment-240218434 I believe that appiles to CURRENT_ROW as well. I'll do that.

**Comment by maciejgrzybek** [comment]
Looks good. Thanks for adding assertion (https://github.com/Teradata/presto/pull/304#issuecomment-240218930).

**Comment by ebd2** [comment]
Closing to open FB PR.
